### ISSUE TITLE
Built-in support for `some` / `every` quantifiers

### ISSUE THREAD

**Issue by rtheunissen** [cue-lang/cue#2947]
A constraint pattern that we are using very often is to check that every value in a list matches a predicate, or that at least one value matches a predicate. Effectively [some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) and [every](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every). 

Currently we are implementing this in Cue using `len` checks on list comprehensions:

```
// ALL
if len([for index, value in somelist if { predicate } {_}]) == len(somelist) { ... }

// NONE
if len([for index, value in somelist if { predicate } {_}]) == 0 { ... }

// SOME
if len([for index, value in somelist if { predicate } {_}]) > 0 { ... }
```

This is not ideal for two main reasons:
- It is intensely verbose especially when multiple of these are nested (below)
- It requires that we evaluate the predicate for every item regardless

```
if {
  len([for v1 in list_a if {
    (len([for v2 in v1.list_b if {
      (len([for v3 in v2.record_d if {
        len([for v4 in v3.list_e if {
          _v4 < 0
        } {_}]) > 0
      } {_}]) > 0)
    } {_}]) == len(v1.list_b))
  } {_}]) > 0
} {
    ...
}
```

🤮 

**Describe the solution you'd like**

I would like to have built-in list functions that provide this functionality. Something that takes a list and a predicate. 

For example, Pkl provides these methods for lists:
- https://pkl-lang.org/package-docs/pkl/0.25.2/base/List#every
- https://pkl-lang.org/package-docs/pkl/0.25.2/base/List#any

**Describe alternatives you've considered**

The `len` with list comprehensions is the only way that we've been able to model this. If there is a better way to do this in Cue currently we would love to know about it.

**Comment by nagl-resourcely** [comment]
There isn't, but we've always intended to break ground on cue builtins. This would be a great excuse to do so.

**Comment by noamtd-msft** [comment]
I think that for the "every" case, this can be handled a bit more succinctly like this:

```cuelang
someList: [
    1,
    3,
    4,
    -2
]

#predicate: >0
allNumsPositive: (someList & [...#predicate]) != _|_
```

Although I'm not sure if this is robust enough to handle every use case

**Comment by nagl-resourcely** [comment]
That works for all. I'm working on a `some` idiom... we may be able to dry this out with the cue function pattern.

**Comment by NoamTD** [comment]
yeah, some is a bit trickier - it can be handled by negating the predicate and checking for bottom:

```cuelang
someList: [
    -5,
    7,
    -1
]

#nonPositive: <=0
someNumsPositive: (someList & [...#nonPositive]) == _|_
```

and the intention would be made clearer with `isconcrete` and `not` builtins - but as mentioned in the issue, this is most likely less efficient than some built-in that the language could provide, with short  circuiting

**Comment by jpluscplusm** [comment]
To my eye, `every` and `some` can now be encoded using the built-in `matchN` validator, which is detailed here: https://cuelang.org/docs/howto/use-the-built-in-function-matchn/. There's also a short section demonstrating its use [with composite values](https://cuelang.org/docs/howto/use-the-built-in-function-matchn/#composite-data-structures).

@rtheunissen Does it achieve what you need?

**Comment by mvdan** [comment]
I believe @jpluscplusm is correct; this can now be done with `matchN`. He didn't get a response saying otherwise, so I'm going to close this as fixed for now.
